### ISSUE TITLE
Shipping extension in RPM format

### ISSUE THREAD

**Issue by fiftydinar** [p91paul/middleclickclose#46]
EDIT:

I removed the RPM that I hosted, as I use system-wide `gnome-extensions` BlueBuild module to install extensions now.
It's much easier to maintain this than fiddling with RPM & COPR.
https://blue-build.org/reference/modules/gnome-extensions/

I wanted to make Gnome extensions that I used, which installed system-wide for custom OS images based on Fedora.

So I originally made RPM because there was no other way to achieve what I want at the time I made the issue.

Another issue was the appliance as a Fedora packager, which is even more tedious process than making simple RPMs imo.

For anyone who wants to help on this instead of me, feel free to do so.

Here's the solid spec file example if anyone wants to continue this effort:
```
%global extension   middleclickclose
%global uuid        %{extension}@paolo.tranquilli.gmail.com
%global gettext     org.gnome.shell.extensions-%{extension}
%global version     30

Name:           gnome-shell-extension-%{extension}
Version:        %{version}
Release:        1%{?dist}
Summary:        Close windows with a button click (the middle one by default) when in overview mode
License:        GPLv2
URL:            https://github.com/p91paul/middleclickclose
Source0:        %{url}/archive/refs/tags/%{version}.tar.gz
BuildArch:      noarch
BuildRequires:  gettext
Requires:       gnome-shell >= 45, gnome-shell < 47

%description
Close windows with a button click (the middle one by default) when in overview mode

%prep
%autosetup -n %{extension}-%{version}

%install
# install main extension files
install -d -m 0755 %{buildroot}%{_datadir}/gnome-shell/extensions/%{uuid}
find src -mindepth 1 -maxdepth 1 ! -path "*po*" ! -path "*schemas*" -exec cp -r {} %{buildroot}%{_datadir}/gnome-shell/extensions/%{uuid}/ \;

# install the schema file
install -D -p -m 0644 \
    src/schemas/org.gnome.shell.extensions.%{extension}.gschema.xml \
    %{buildroot}%{_datadir}/glib-2.0/schemas/org.gnome.shell.extensions.%{extension}.gschema.xml

# install locale files
pushd src/po
for po in *.po; do
    install -d -m 0755 %{buildroot}%{_datadir}/locale/${po%.po}/LC_MESSAGES
    msgfmt -o %{buildroot}%{_datadir}/locale/${po%.po}/LC_MESSAGES/%{gettext}.mo $po
done
popd
%find_lang %{gettext}

%files -f %{gettext}.lang
%doc README.md
%license LICENSE
%{_datadir}/glib-2.0/schemas/*.gschema.xml
%{_datadir}/gnome-shell/extensions/%{uuid}/

%changelog
* Sat Apr 06 2024 Your Name <yourname@mail.com> - 30
- improvement: Added Gnome 46 support
```

**Comment by dreua** [comment]
Thanks for sharing! Packaging for Fedora is also somewhere on my todo list ;)

I am already a Fedora packager so that might be one step easier for me, but you could become one yourself in the process which is also an opportunity, so let me know which way you would like to proceed.

BTW you posted the wrong link to open weather ;)

**Comment by fiftydinar** [comment]
> Thanks for sharing! Packaging for Fedora is also somewhere on my todo list ;)
> 
> I am already a Fedora packager so that might be one step easier for me, but you could become one yourself in the process which is also an opportunity, so let me know which way you would like to proceed.
> 
> BTW you posted the wrong link to open weather ;)

Link fixed. :)

Thanks for the opportunity, I will take a research on it, to see If it's suited for me or not. Just taking a glance on requirements, I don't like bugzilla's interface, so it will take some time to get used. And I generally, don't like mailing lists.

Can there be more Fedora packagers for the same package?
If not, I would have no problem you taking the responsibility. Meanwhile, I can help from the side.

One thing I have in mind for spec file, is future-proofing the "installing main extension files" part, by copying everything but "po" & "schemas" folder, so I don't have to manually include new .js file. Don't know how to achieve that currently, so some assistance there would be nice.

**Comment by fiftydinar** [comment]
> Yes, you can become comaintainer once it is accepted, you'll just need to create a Fedora Account (fas) afaik.
> 
> Have you tried shell globbing, with *.js for example? cp or the shell around it should be able to handle that. Maybe even advanced globbing like `src/*.{js,json,whateverExtension}` should work. Assuming you know all possible extensions in advance.

That would be awesome, I already have Fedora Account!

Just tried your suggestion & it works wonders! Github & COPR is updated with your changes. I know that other, much bigger extensions, usually ship some .css files & some icons if needed, but I don't think that this extension needs that at all, as it's light in size.

**Comment by gtsiam** [comment]
You seem to essentially be re-implementing `make pack`, which creates the package we'll upload to gnome-extensions. I'd suggest you run that and then unzip the resulting file wherever you want to avoid issues during your rpm build.

Also the rpm spec file should probably not be in the same commit history as the main project. Arch PKGBUILDs can download sources from the internet. I'm assuming RPM specs can do the same? Alternatively, consider using git submodules.

I'm going to close this for now, but feel free to make a pull request mentioning this in the readme (I will nitpick if you reference the master branch though, so wait until after the gnome 45 release for proper versioning - soon™). FYI, most projects tag and do github releases on new versions for packaging if you want to stay up-to-date.

**Comment by fiftydinar** [comment]
> You seem to essentially be re-implementing `make pack`, which creates the package we'll upload to gnome-extensions. I'd suggest you run that and then unzip the resulting file wherever you want to avoid issues during your rpm build.
> 
> Also the rpm spec file should probably not be in the same commit history as the main project. Arch PKGBUILDs can download sources from the internet. I'm assuming RPM specs can do the same? Alternatively, consider using git submodules.
> 
> I'm going to close this for now, but feel free to make a pull request mentioning this in the readme (I will nitpick if you reference the master branch though, so wait until after the gnome 45 release for proper versioning - soon™). FYI, most projects tag and do github releases on new versions for packaging if you want to stay up-to-date.

I appreciate the feedback.
I'm aware of some of the remarks you made, as it was customized for my needs only, as mentioned in my 1st post.

I will try your suggestion with make pack.

I would not use master branch officially of course, I would use normal release versioning like you mentioned.

As far as I'm informed, when you apply for official Fedora packager, spec file & it's commits are hosted on src.fedoraproject.org, so it's separate from this repo, unlike my current implementation.

I would not want to make RPM official, where users would need to have RPMFusion installed, so I want it to be from official Fedora repos.

That means, I would either need to apply to be a Fedora packager, or someone else who is interested in this, in this case @dreua. It's something we need to communicate about. Current state, as speculated about, is for @dreua to be a package maintainer & I would be a co-maintainer. It is not official, there is still some time left for a discussion & final statement to be made. But when this stuff is solved, I believe that every your remark will also be satisfied.

**Comment by dreua** [comment]
Yes, for the Fedora package there will be a separate repository just for the spec file (and possibly auxiliary files) on src.fedoraproject.org and it will download the specified version from this Github repository.

Some projects do keep a spec file (and maybe also packaging files for debian etc.) in their repository in order to do nightly or vendor builds, so that is not unheard of but not necessary either.

I've begun working on the packaging, for review I'd like to have a version that is compatible with Gnome 45 (reviewers may demand it) otherwise I can only package the current `1.9.1` for Fedora <= 38.

Shall we keep that issue open as long as it is worked on? Feel free to assign me.

**Comment by gtsiam** [comment]
See #50 for Gnome 45 support on EGO.

I'd rather tag after the EGO review because they might request changes. Then again, depends on how soon you want to repackage. I could make an rc release tomorrow, but I kind of doubt you can package that.

**Comment by gtsiam** [comment]
@dreua `v29` is now up on EGO and tagged on github. You'll find it under the `v29` tag (I don't see a point in doing semver as we were doing before, plus this change wouldn't need an epoch bump so I only see it as a positive).

**Comment by gtsiam** [comment]
I just noticed the edit:
> I removed the RPM that I hosted, as I use system-wide gnome-extensions BlueBuild module to install extensions now.
It's much easier to maintain this than fiddling with RPM & COPR.
https://blue-build.org/reference/modules/gnome-extensions/

Then this seems fair to close. If anyone is interesting in picking this up, feel free to open a pull request linking to your work in the README.
